c/tskit/tables.c

```diff
@@ -8105,6 +8105,285 @@ tsk_table_collection_subset(
+    bool check_shared_portion = !(options & TSK_NO_CHECK_SHARED);
+
+    ret = tsk_table_collection_init(&self_copy, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_init(&other_copy, 0);
+    if (ret != 0) {
+        goto out;
+    }
+
+    /* node_mapping relates id in other (index) to equivalent node in self or
+     * TSK_NULL if not present */
+    if (other_num_nodes != (tsk_size_t) other->nodes.num_rows) {
+        ret = TSK_ERR_LEN_MAP;
+        goto out;
+    }
+
+    for (k = 0; k < (tsk_id_t) other_num_nodes; k++) {
+        if (node_mapping[k] >= (tsk_id_t) self->nodes.num_rows) {
+            ret = TSK_ERR_BAD_MAP;
+            goto out;
+        }
+        if (node_mapping[k] != TSK_NULL)
+            num_shared_nodes++;
+    }
+
+    self_nodes = malloc(num_shared_nodes * sizeof(*self_nodes));
+    other_nodes = malloc(num_shared_nodes * sizeof(*other_nodes));
+    if (self_nodes == NULL || other_nodes == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    i = 0;
+    for (k = 0; k < (tsk_id_t) other_num_nodes; k++) {
+        if (node_mapping[k] != TSK_NULL) {
+            self_nodes[i] = node_mapping[k];
+            other_nodes[i] = k;
+            i++;
+        }
+    }
+
+    // checking shared portion
+    // TODO: strict sort before checking equality
+    if (check_shared_portion) {
+        ret = tsk_table_collection_copy(self, &self_copy, TSK_NO_INIT);
+        if (ret != 0) {
+            goto out;
+        }
+        ret = tsk_table_collection_copy(other, &other_copy, TSK_NO_INIT);
+        if (ret != 0) {
+            goto out;
+        }
+        ret = tsk_provenance_table_clear(&other_copy.provenances);
+        if (ret != 0) {
+            goto out;
+        }
+        ret = tsk_provenance_table_clear(&self_copy.provenances);
+        if (ret != 0) {
+            goto out;
+        }
+        ret = tsk_table_collection_subset(&self_copy, self_nodes, num_shared_nodes);
+        if (ret != 0) {
+            goto out;
+        }
+        ret = tsk_table_collection_subset(&other_copy, other_nodes, num_shared_nodes);
+        if (ret != 0) {
+            goto out;
+        }
+        ret = !tsk_table_collection_equals(&self_copy, &other_copy);
+        if (ret != 0) {
+            ret = TSK_ERR_DIFF_HISTORIES;
+            goto out;
+        }
+    }
+
+out:
+    tsk_table_collection_free(&self_copy);
+    tsk_table_collection_free(&other_copy);
+    tsk_safe_free(other_nodes);
+    tsk_safe_free(self_nodes);
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_table_collection_union(tsk_table_collection_t *self, tsk_table_collection_t *other,
+    tsk_id_t *node_mapping, tsk_size_t num_nodes, tsk_flags_t options)
+{
+    int ret = 0;
+    tsk_id_t k, i, new_ind, new_pop, new_parent, new_child;
+    tsk_node_t node;
+    tsk_individual_t ind;
+    tsk_population_t pop;
+    tsk_edge_t edge;
+    tsk_mutation_t mut;
+    tsk_site_t site;
+    tsk_id_t *X_node_map = NULL;
```

Comment:
why the `X_`?

Reply:
because I already used node_map for sth else. you suggested this name, but I also don't know where the X comes from heheh

Reply:
oh, nvm I changed the other one to node_mapping.